device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -1220,10 +1214,10 @@ private void addToCallbackQueue(IotHubTransportPacket packet)
         //Codes_SRS_IOTHUBTRANSPORT_28_002: [This function shall add the packet to the callback queue if it has a callback.]
         if (packet.getCallback() != null)
         {
+            this.callbackPacketsQueue.add(packet);
```

Comment:
Deliberately adding the packet outside of the synchronized block. Doing so here and below fixes the deadlock that we've been seeing. These queues are concurrent queues and may block on add/remove if another thread is using them. For safety, we don't want this queue addition to be done inside of another synchronization block as that may lead to deadlock